packages/backend-common/src/database/SingleConnection.test.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ConfigReader } from '@backstage/config';
+import { createDatabaseClient } from './connection';
+import { SingleConnectionDatabaseManager } from './SingleConnection';
+
+jest.mock('./connection');
+
+describe('SingleConnectionDatabaseManager', () => {
+  const createConfig = (data: any) =>
+    ConfigReader.fromConfigs([
+      {
+        context: '',
+        data,
+      },
+    ]);
+
+  const defaultConfigOptions = {
+    backend: {
+      database: {
+        client: 'pg',
+        connection: {
+          host: 'localhost',
+          user: 'foo',
+          password: 'bar',
+          database: 'foodb',
+        },
+      },
+    },
+  };
+  const defaultConfig = () => createConfig(defaultConfigOptions);
+
+  // This is similar to the ts-jest `mocked` helper.
+  const mocked = (f: Function) => f as jest.Mock;
+
+  describe('SingleConnectionDatabaseManager.fromConfig', () => {
+    it('accesses the backend.database key', () => {
+      const getConfig = jest.fn();
+      const config = defaultConfig();
+      config.getConfig = getConfig;
+
+      SingleConnectionDatabaseManager.fromConfig(config);
+
+      expect(getConfig.mock.calls[0][0]).toEqual('backend.database');
+    });
+  });
+
+  describe('SingleConnectionDatabaseManager.forPlugin', () => {
+    const manager = SingleConnectionDatabaseManager.fromConfig(defaultConfig());
+
+    it('connects to a database scoped to the plugin', async () => {
+      const pluginId = 'test1';
+      await manager.forPlugin(pluginId).getClient();
+
+      expect(mocked(createDatabaseClient).mock.calls).toHaveLength(1);
+      const mockCalls = mocked(createDatabaseClient).mock.calls.splice(-1);
```

Comment:
Could use `.toHaveBeenCalledWith` here and in other places. `.toHaveBeenLastCalledWith` and `.toHaveBeenNthCalledWith` for the cases with multiple calls

Reply:
I observe that the mock calls don't get reset after every test, so this ensures that the calls are properly reset to 0 after every test run and that the tests need not run in order.

Unless there's something I'm missing here?

Reply:
That's expected - mocks get created at the level the mock happens and exist for the lifetime of the mock. You may want to add `afterEach(() => jest.resetAllMocks())` inside the top level `describe`s. This keeps the mocks but reset their internal state after each individual test.

Reply:
OK I didn't know that existed. Let me fix.

Reply:
toHaveBeenCalledWith makes matching against an object difficult. The Config object requires a call to get() first:
```
      expect(mocked(createDatabaseClient)).toHaveBeenCalledWith(
        WhatMatcherFitsHere??,
        expect.objectContaining({
          connection: {
            database: `backstage_plugin_${pluginId}`
          }
        })
      )
```

and the other test also checks that the provided databases are different - I'm not explicitly check that the databases match some fixed constant.

```
      const plugin1CallArgs = mockCalls[0];
      const plugin2CallArgs = mockCalls[1];
      expect(plugin1CallArgs[1].connection.database).not.toEqual(
        plugin2CallArgs[1].connection.database,
      );
```

I've changed to `toHaveBeenCalledTimes` though.

Reply:
Yeah the first one is a bit tricky, could keep as is. Another option would be to actually use a mocked implementation and assert the return value from that instead, basically `createDatabaseClient.mockImplementation(config, overrides) => ({ config, overrides })` and then check the return value of `getClient`. You'd likely have to throw some `any` in there for that to work though.

For the second one I'd actually prefer the hardcoded values, as there are many ways that broken behaviour could still lead to a passing test, and we should know exactly what database names to expect right?